Wrap Coil:Cooling:DX:SingleSpeed:ThermalStorage

joseph-robertson opened the pull request:
Pull request overview
---------------------

Companion PR: https://github.com/NREL/OpenStudio/pull/5059

Link to relevant GitHub Issue(s) if appropriate:

Link to the **Ubuntu 20.04 .deb** installer to use for CI Testing. If not set, it will default to latest official release.
[OpenStudio Installer]: https://github.com/NREL/OpenStudio/releases/download/v3.8.0-rc2/OpenStudio-3.8.0-rc2+af997f4fef-Ubuntu-20.04-x86_64.deb


This Pull Request is concerning:

 - [x] **Case 1 - `NewTest`:** a new test for a new model API class,

----------------------------------------------------------------------------------------------------------

### Case 1: New test for a new model API class

Please include which class(es) you are adding a test to specifically test for.
Include a link to the OpenStudio Pull Request in which you are adding the new classes, or the class itself if already on develop.
<!---
> eg:
>
> This pull request is in relation with the Pull Request [NREL/OpenStudio#3031](https://github.com/NREL/OpenStudio/pull/3031), and  will specifically test for the following classes:
> * `AirTerminalSingleDuctConstantVolumeFourPipeBeam`
> * `CoilCoolingFourPipeBeam`
> * `CoilHeatingFourPipeBeam`
> * Additionally it explicitly tests for the existing class [TableMultiVariableLookUp](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/TableMultiVariableLookup.hpp)
-->

#### Work Checklist

The following has been checked to ensure compliance with the guidelines:

 - [ ] Tests pass either:
     - [ ] with official OpenStudio release (include version):
         - [ ] A matching OSM test has been added from the successful run of the Ruby one with the official OpenStudio release
         - [ ] The label `AddedOSM` has been added to this PR
         - [ ] All new `out.osw` have been committed

     - [ ] with current develop (incude SHA):
         - [x] The label `PendingOSM` has been added to this PR
         - [x] A matching OSM test has not yet been added because the official release is pending, but `model_tests.rb` has a TODO.
            ```ruby
            def test_airterminal_cooledbeam_rb
              result = sim_test('airterminal_cooledbeam.rb')
            end

            # TODO: To be added in the next official release after: 2.5.0
            # def test_airterminal_fourpipebeam_osm
            #   result = sim_test('airterminal_fourpipebeam.osm')
            # end
            ```
        - [ ] No `out.osw` have been committed as they need to be run with an official OpenStudio version


 - [ ] **Ruby test is stable**: when run multiple times on the same machine, it produces the same total site kBTU.
    Please paste the heatmap png generated after running the following commands:
     - [ ] I ensured that I assign systems/loads/etc in a repeatable manner (eg: if I assign stuff to thermalZones, I do `model.getThermalZones.sort_by{|z| z.name.to_s}.each do ...` so I am sure I put the same ZoneHVAC systems to the same zones regardless of their order)
     - [ ] I tested stability using `process_results.py` (see `python process_results.py --help` for usage).
    Please paste the text output or heatmap png generated after running the following commands:
        ```bash
        # Clean up all custom-tagged OSWs
        python process_results.py test-stability clean
        # Run your test 5 times in a row. Replace `testname_rb` (eg `airterminal_fourpipebeam_rb`)
        python process_results.py test-stability run -n testname_rb
        # Check that they all passed
        python process_results.py test-status --tagged
        # Check site kBTU differences
        python process_results.py heatmap --tagged

        ```

 - [x] **Object has been added to `autosize_hvac.rb` to ensure the autosizedXXX values methods do work**

----------------------------------------------------------------------------------------------------------

### Review Checklist

 - [ ] Code style (indentation, variable names, strip trailing spaces)
 - [ ] Functional code review (it has to work!)
 - [ ] Matching OSM test has been added or `# TODO` added to `model_tests.rb`
 - [ ] Appropriate `out.osw` have been committed
 - [ ] Test is stable
 - [ ] Object is tested in `autosize_hvac` as appropriate
 - [x] The appropriate labels have been added to this PR:
   - [x] One of: `NewTest`, `TestFix`, `NewTestForExisting`, `Other`
   - [x] If `NewTest`: add `PendingOSM` or `AddedOSM`

joseph-robertson committed:
Add new simulation test for cooling dx thermal storage.

joseph-robertson committed:
Call new test from model_tests.rb.

Commit:
Commit rubocop --auto-correct (Ruby ruby 2.7.8p225 (2023-03-30 revision 1f4d455848) [x86_64-linux], Rubocop 0.81.0)

joseph-robertson committed:
Update triquadratic and table lookup curves.

Commit:
Commit rubocop --auto-correct (Ruby ruby 2.7.8p225 (2023-03-30 revision 1f4d455848) [x86_64-linux], Rubocop 0.81.0)

joseph-robertson committed:
Add the ems controls.

joseph-robertson committed:
Merge branch 'coil-thermal-storage' of github.com:NREL/OpenStudio-resources into coil-thermal-storage

Commit:
Commit rubocop --auto-correct (Ruby ruby 2.7.8p225 (2023-03-30 revision 1f4d455848) [x86_64-linux], Rubocop 0.81.0)

joseph-robertson committed:
Change all autocalculate to autosize.

joseph-robertson committed:
Minor change to trigger ci.

jmarrec reviewed (changes requested):
Too big to check in details, but I assume this mimics the RetailTES IDF Example so fine by me. I'd have liked to see it tested in all of the possible containingZoneHVAC / containingHVAC components but this is complicated enough that I don't really care, provided you do add C++ ForwardTranslator tests to check the node connections are fine.

You do need to add it to **autosize_hvac** though since it has autosizable fields (this will check the SQL methods autosizedXXX)

joseph-robertson commented:
> Too big to check in details, but I assume this mimics the RetailTES IDF Example so fine by me. I'd have liked to see it tested in all of the possible containingZoneHVAC / containingHVAC components but this is complicated enough that I don't really care, provided you do add C++ ForwardTranslator tests to check the node connections are fine.
> 
> You do need to add it to **autosize_hvac** though since it has autosizable fields (this will check the SQL methods autosizedXXX)

Yes, the idea was to mimic RetailPackagedTESCoil.idf.

joseph-robertson committed:
Start adding ts coil in autosize_hvac.

joseph-robertson committed:
Explicitly call out all autosizable fields.

joseph-robertson committed:
Also hardset fluid storage tank rating temperature.

joseph-robertson committed:
Stub setters we may need in autosize_hvac test.

jmarrec committed:
Merge branch 'develop' into coil-thermal-storage

jmarrec commented on model/simulationtests/autosize_hvac.rb:
```diff
@@ -831,6 +831,38 @@ def curve_quadratic(model, c_1constant, c_2x, c_3xPOW2, minx, maxx, miny, maxy)
 humidistat.setHumidifyingRelativeHumiditySetpointSchedule(dehumidify_sch)
 zones[40].setZoneControlHumidistat(humidistat)
 
+air_loop_unitary = OpenStudio::Model::AirLoopHVACUnitarySystem.new(m)
+coil = OpenStudio::Model::CoilCoolingDXSingleSpeedThermalStorage.new(m)
```
undefined variable m

jmarrec commented on model/simulationtests/autosize_hvac.rb:
```diff
@@ -831,6 +831,38 @@ def curve_quadratic(model, c_1constant, c_2x, c_3xPOW2, minx, maxx, miny, maxy)
+air_loop_unitary.addToNode(air_loop.supplyInletNode)
+air_loop_unitary.setControllingZoneorThermostatLocation(zones[43])
+atu = OpenStudio::Model::AirTerminalSingleDuctConstantVolumeNoReheat.new(m, s1)
+air_loop.addBranchForZone(z, atu)
```
undefined z

jmarrec committed:
Ignore DistrictHeating:Steam from autosize_hvac

jmarrec committed:
Use proper variables in autosize_hvac: the test fails though!

```
   ** Severe  ** Node NODE 179 connects to no component
   **   ~~~   ** Occurs in Supply Air Path=AIR LOOP NODE 177 SUPPLY PATH
   **   ~~~   ** Check the connection to a ZoneHVAC:EquipmentConnections object
   **   ~~~   ** Check if this component is missing from the Supply Air Path
   ** Severe  ** An outlet node in AirLoopHVAC="AIR LOOP" is not connected to any zone
   **   ~~~   ** Could not match ZoneEquipGroup Inlet Node="NODE 177" to any Supply Air Path or controlled zone
   **  Fatal  ** Preceding errors cause termination
```

jmarrec commented:
I made adjustments to autosize_hvac to use the proper variables. The tests fails though

```
   ** Severe  ** Node NODE 179 connects to no component
   **   ~~~   ** Occurs in Supply Air Path=AIR LOOP NODE 177 SUPPLY PATH
   **   ~~~   ** Check the connection to a ZoneHVAC:EquipmentConnections object
   **   ~~~   ** Check if this component is missing from the Supply Air Path
   ** Severe  ** An outlet node in AirLoopHVAC="AIR LOOP" is not connected to any zone
   **   ~~~   ** Could not match ZoneEquipGroup Inlet Node="NODE 177" to any Supply Air Path or controlled zone
   **  Fatal  ** Preceding errors cause termination
```

jmarrec commented:
Running the test_coil_cooling_dx_singlespeed_thermalstorage_rb

```
$ cat testruns/coil_cooling_dx_singlespeed_thermalstorage.rb/run/eplusout.err
```

extract, redacted some useless stuff:

```
Program Version,EnergyPlus, Version 24.1.0-9d7789a3ac, YMD=2024.05.07 09:45,
   ************* Beginning Zone Sizing Calculations
[...]
   ************* Beginning Simulation
[...]
   ** Warning ** Enthalpy out of range (PsyTsatFnHPb)
   **   ~~~   **  Routine=CalcTESCoilCoolingOnlyMode, Environment=RUN PERIOD 1, at Simulation time=01/16 14:00 - 14:15
   **   ~~~   **  Enthalpy=-749753.06945 Pressure= 99200.00
   **   ~~~   **  Initial Resultant Temperature= -59.98
   ** Warning ** Temperature out of range [-100. to 200.] (PsyPsatFnTemp)
   **   ~~~   **  Routine=CalcTESCoilCoolingOnlyMode, Environment=RUN PERIOD 1, at Simulation time=01/16 14:00 - 14:15
   **   ~~~   **  Input Temperature=-762.54
   ** Warning ** Calculated Humidity Ratio invalid (PsyWFnTdbH)
   **   ~~~   **  Routine=CalcTESCoilCoolingOnlyMode Environment=RUN PERIOD 1, at Simulation time=01/23 14:00 - 14:15
   **   ~~~   **  Dry-Bulb= -59.98 Enthalpy= -587592.367
   **   ~~~   ** Calculated Humidity Ratio= -0.2206 ... Humidity Ratio set to .00001
   ** Warning ** Coil control failed for AirLoopHVAC:UnitarySystem:AIR LOOP HVAC UNITARY SYSTEM 1
   **   ~~~   **   sensible part-load ratio determined to be outside the range of 0-1.
   **   ~~~   ** Sensible load to be met = -5.96 (watts), and the simulation continues.
   **   ~~~   **  Environment=RUN PERIOD 1, at Simulation time=09/18 09:00 - 09:15

   ************* Simulation Error Summary *************

   ************* ===== Recurring Error Summary =====
   ************* The following recurring error messages occurred.
   *************
   *************  ** Warning ** Enthalpy out of range (PsyTsatFnHPb)
   *************  **   ~~~   **   This error occurred 18053 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=-42406.646858 J/kg  Min=-378342195.011322 J/kg
   *************
   *************  ** Warning ** Temperature out of range [-100. to 200.] (PsyPsatFnTemp)
   *************  **   ~~~   **   This error occurred 25011 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=-100.013180 C  Min=-376530.671875 C
   *************
   *************  ** Warning ** Calculated Humidity Ratio invalid (PsyWFnTdbH)
   *************  **   ~~~   **   This error occurred 979 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=-0.000374 []  Min=-5.789433 []
   *************
   *************  ** Warning ** AirLoopHVAC:UnitarySystem "AIR LOOP HVAC UNITARY SYSTEM 1" - sensible part-load ratio out of range error continues. Sensible load statistics:
   *************  **   ~~~   **   This error occurred 2 total times;
   *************  **   ~~~   **   during Warmup 0 times;
   *************  **   ~~~   **   during Sizing 0 times.
   *************  **   ~~~   **   Max=-5.966155  Min=-5.966155
   *************
   *************
   ************* ===== Final Error Summary =====

   ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
   ************* EnergyPlus Sizing Error Summary. During Sizing: 7 Warning; 0 Severe Errors.
   ************* EnergyPlus Completed Successfully-- 44063 Warning; 0 Severe Errors; Elapsed Time=00hr 00min  3.33sec
```

jmarrec commented:
> I made adjustments to autosize_hvac to use the proper variables. The tests fails though
> 
> ```
>    ** Severe  ** Node NODE 179 connects to no component
>    **   ~~~   ** Occurs in Supply Air Path=AIR LOOP NODE 177 SUPPLY PATH
>    **   ~~~   ** Check the connection to a ZoneHVAC:EquipmentConnections object
>    **   ~~~   ** Check if this component is missing from the Supply Air Path
>    ** Severe  ** An outlet node in AirLoopHVAC="AIR LOOP" is not connected to any zone
>    **   ~~~   ** Could not match ZoneEquipGroup Inlet Node="NODE 177" to any Supply Air Path or controlled zone
>    **  Fatal  ** Preceding errors cause termination
> ```

wow it took me way too long to find why, because I also had E+ crashing on me if I didn't add Cooling Only curves, but this is because you override the `air_loop` variable, so the loop below that add terminal to it was actually connecting to yours and the original `air_loop`  was connected to nothing

jmarrec committed:
Fix the autosize test for the TES coil. It's awful but it works

jmarrec committed:
Add a Python version of the test

Commit:
Commit rubocop --auto-correct (Ruby ruby 2.7.8p225 (2023-03-30 revision 1f4d455848) [x86_64-linux], Rubocop 0.81.0)

jmarrec commented:
I had to make a fix to the OS SDK side. 

* https://github.com/NREL/OpenStudio/pull/5187

I fixed the autosize hvac, and I added a python version of the test

jmarrec commented:
@joseph-robertson I'm maybe preemptively merging this, but I'm tired to look at this. It works locally.

The code changes I made are burning my eyes a bit, but they work.

Feel free to review and do adjustments in another PR if you want.

jmarrec commented on model/simulationtests/autosize_hvac.rb:
```diff
@@ -81,17 +81,96 @@ def curve_biquadratic(model, c_1constant, c_2x, c_3xPOW2, c_4y, c_5yPOW2, c_6xTI
+  constant_bivariate = OpenStudio::Model::CurveBiquadratic.new(model)
+  constant_bivariate.setName('Constant Bivariate Curve')
+  constant_bivariate.setCoefficient1Constant(1.0)
+  constant_bivariate.setMinimumValueofx(-100.0)
+  constant_bivariate.setMaximumValueofx(100.0)
+  constant_bivariate.setMinimumValueofy(-100.0)
+  constant_bivariate.setMaximumValueofy(100.0)
+
+  constant_trivariate = OpenStudio::Model::CurveTriquadratic.new(model)
+  constant_trivariate.setCoefficient1Constant(1.0)
+  constant_trivariate.setMinimumValueofx(-100.0)
+  constant_trivariate.setMaximumValueofx(100.0)
+  constant_trivariate.setMinimumValueofy(-100.0)
+  constant_trivariate.setMaximumValueofy(100.0)
+  constant_trivariate.setMinimumValueofz(-100.0)
+  constant_trivariate.setMaximumValueofz(100.0)
+
+  object_list_to_curve = {
+    'UnivariateFunctions' => constant_univariate,
+    'BivariateFunctions' => constant_bivariate,
+    'TrivariateFunctions' => constant_trivariate
+  }
+
+  # Scan the curve fields, assign a constant curve with the right number of
+  # independent variables
+  iddObject = tes_coil.iddObject
+  iddObject.numFields.times do |i|
+    field = iddObject.getField(i).get
+    field_name = field.name
+    next unless field_name.include?('Curve')
+
+    props = field.properties
+    olist = props.objectLists.first
+    setter = 'set' + field_name.gsub(' ', '')
+    raise "Undefined method #{setter} for field_name=#{field_name} at index #{i}" unless tes_coil.respond_to?(setter)
+
+    tes_coil.send(setter, object_list_to_curve[olist])
+  end
+
+  tes_coil.setCondenserType('EvaporativelyCooled')
```
Yeah, I know. Not great as a far as readability goes, but the point of this test is to check the autosized SQL calls, not making a clean / physically representative model

jmarrec commented on model/simulationtests/autosize_hvac.rb:
```diff
@@ -831,6 +910,20 @@ def curve_quadratic(model, c_1constant, c_2x, c_3xPOW2, minx, maxx, miny, maxy)
 humidistat.setHumidifyingRelativeHumiditySetpointSchedule(dehumidify_sch)
 zones[40].setZoneControlHumidistat(humidistat)
 
+# TES Coil CoilCoolingDXSingleSpeedThermalStorage
+tes_air_loop_unitary = OpenStudio::Model::AirLoopHVACUnitarySystem.new(model)
+tes_coil = make_tes_coil(model)
+fan = OpenStudio::Model::FanOnOff.new(model)
+tes_air_loop_unitary.setAvailabilitySchedule(model.alwaysOnDiscreteSchedule)
+tes_air_loop_unitary.setCoolingCoil(tes_coil)
+tes_air_loop_unitary.setSupplyFan(fan)
+tes_air_loop = OpenStudio::Model::AirLoopHVAC.new(model)
+tes_air_loop_unitary.addToNode(tes_air_loop.supplyInletNode)
+tes_air_loop_unitary.setControllingZoneorThermostatLocation(zones[43])
+tes_atu = OpenStudio::Model::AirTerminalSingleDuctConstantVolumeNoReheat.new(model, s1)
+tes_air_loop.addBranchForZone(zones[43], tes_atu)
+tes_air_loop.setName('AirLoop with TES Coil')
```
Renamed the variables to be tes_ prefixed so I don't steal others defined before and normally used after (air_loop in particular).

jmarrec commented on model/simulationtests/autosize_hvac.rb:
```diff
@@ -1090,10 +1183,13 @@ def curve_quadratic(model, c_1constant, c_2x, c_3xPOW2, minx, maxx, miny, maxy)
 
-  when 26, 27, 28, 29, 30, 31, 32, 33, 38, 40
+  when 26, 27, 28, 29, 30, 31, 32, 33, 38, 40, 43
     # Previously used for the unitary systems, dehum, etc
-  else
+  when 0, 44
+    # This wasn't assigned yet and is for grabs
     puts "Nothing added to #{zn.name}, index #{zone_index}"
+  else
+    raise "Nothing added to #{zn.name}, index #{zone_index}"
```
Adjust here, to force people keeping it up to date
